WIP: Finish Ethereum Wire Protocol implementation

preston-evans98 opened the pull request:
Implement each message type.
Implement some tests cases from EIP 2481

preston-evans98 commented on src/downloader/message_decoder.rs:
```diff
@@ -63,4 +88,284 @@ mod tests {
+    }
+
+    #[test]
+    fn decode_receipts() {
+        let expected_bytes =
+        hex!("f90172820457f9016cf90169f901668001b90100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ...
+        let result = decode_rlp_message(EthMessageId::Receipts, &expected_bytes);
+        let some_message = result.unwrap();
```
This decode operation fails for reasons that aren't entirely clear to me. The observed error type is `RlpExpectedToBeList`.  I suspect that the problem may be in the RlpDerive macro, since `decode_node_data` also fails and both of those types have entries of the form `Vec<Vec<T>>`, but I don't have any solid leads.

preston-evans98 replied:
Update: the problem has to do with the definition of `Receipt` in the [ethereum crate](https://docs.rs/ethereum/0.9.0/ethereum/struct.Receipt.html). An RLP encoded receipt from the crate represents the `state_root` as a 32 byte list, while the EIP 2481 test case represents it as a single zero byte. 

I believe the fix is to change the definition of `ethereum::Receipt` to:
```
pub struct Receipt {
	pub state_root: U256, // This used to be H256
	pub used_gas: U256,
	pub logs_bloom: Bloom,
	pub logs: Vec<Log>,
}
```

preston-evans98 replied:
Confirmed: Making this change locally causes the test to pass. I've opened an [issue](https://github.com/rust-blockchain/ethereum/issues/19) on the `ethereum` repo and will submit a PR to fix shortly.

battlmonstr replied:
:+1:

vorot93 commented:
@preston-evans98 Hey, thanks for the PR! Is this still a draft you will be working on?

preston-evans98 commented:
I do plan to keep working on it! Unfortunately, I’m traveling at the moment and likely won’t make much progress for a few days. If you’d prefer to merge as is, I’m happy to make it a full PR and submit follow ups to fill out the code coverage and fix the decoding bugs.

vorot93 commented:
Take your time, just make sure to mark it as ready to review when you're done.

preston-evans98 commented:
This commit is ready to merge as soon as https://github.com/rust-blockchain/ethereum/issues/19  is resolved!

battlmonstr reviewed (approved):
LGTM. Awesome tests!

Rjected commented:
Hey, I'm interested in using some of this - what's left to do on this PR? Could I help out in any way?

preston-evans98 commented:
> Hey, I'm interested in using some of this - what's left to do on this PR? Could I help out in any way?

Not much, last I checked! I started a new job and had to get open source contributions approved with legal, so this just fell by the wayside. It is possible that some dependencies have changed though, so there may be some reformatting required. I'll try to take a look tonight.

preston-evans98 commented:
@vorot93 would you mind kicking off the approval workflow when you have a minute?

battlmonstr commented:
>would you mind kicking off the approval workflow when you have a minute?

@preston-evans98 if you rebase onto master, solve conflicts, and then force push, the workflow should be triggered automatically. Currently GH shows that it has some conflicts, so a manual rebase is still needed. Please try to squash and rebase instead of merging.

preston-evans98 commented:
@battlmonstr I've rebased and squashed. I'm not seeing any conflicts on my end but it doesn't look like CI reran automatically. Do you see anything I've missed?
